api/api/utils/strings.py

```diff
@@ -0,0 +1,53 @@
+import re
+from urllib.parse import quote, unquote
+
+
+def convert_grp(grp: str) -> str | None:
+    """
+    Convert a hex value into a character. Return None if the conversion results in
+    a character that cannot be used as a URI component.
+    """
+    try:
+        converted = chr(int(grp, 16))
+        # Decoded strings should be usable as URI components
+        quote(converted)
+        return converted
+    except UnicodeEncodeError:
+        return None
+
+
+def decode_data(data: str | None = "") -> str:
+    if not data:
+        return ""
+
+    regexes = [
+        r"\\(x)([\da-f]{2})",  # \\xe9 - é
+        r"\\(u)([\da-f]{4})",  # \\u00e9 - é
+        r"(u)([\da-f]{4})",  # u00e9 - é
+    ]
```

Comment:
Can these be combined into a single regex?

```r"\\x([\da-f]{2})|\\(x|u)([\da-f]{4})"```

It should theoretically be faster than multiple separate regexes, especially because we're using `re.sub` and also doing string manipulation in the loop.

Can be multiline if needed. I'm not sure which performs better, if it's faster to iterate on multiple smaller regexes or a single larger one.

Either way, we should definitely compile these at the module level, rather than using the `re` functions, otherwise regex compilation is happening (almost[^caching]) every time the `decode_data` function happens.

[^caching]: See note here regarding caching: http://127.0.0.1:33081/python_3/doc/library/re.html#//apple_ref/Function/re.compile. Basically: it isn't a guarantee, and this is inside a tight, O(n) loop, so it's worth compiling these.